File: custom_adapter/AdmobAdapter/Objc/AlxAdmobBaseAdapter.m
//
//  AlxAdmobBaseAdapter.m
//  AlxAdsOCDemo
//

#import "AlxAdmobBaseAdapter.h"
#import <AlxAds/AlxAds-Swift.h>
#import "AlxAdmobMetaInfo.h"
#import "AlxAdmobNetworkExtras.h"

static NSString *const TAG = @"AlxAdmobBaseAdapter";
static NSString *const PARAMETER = @"parameter";

static BOOL _isInitialized = NO;

@implementation AlxAdmobBaseAdapter

#pragma mark - GADMediationAdapter

+ (GADVersionNumber)adapterVersion {
    NSArray<NSString *> *versionComponents = [AlxAdmobMetaInfo.ADAPTER_VERSION componentsSeparatedByString:@"."];
    GADVersionNumber version = {0};
    if (versionComponents.count >= 3) {
        version.majorVersion = [versionComponents[0] integerValue];
        version.minorVersion = [versionComponents[1] integerValue];
        version.patchVersion = [versionComponents[2] integerValue];
    }
    return version;
}

+ (GADVersionNumber)adSDKVersion {
    NSArray<NSString *> *versionComponents = [[AlxSdk getSDKVersion] componentsSeparatedByString:@"."];
    GADVersionNumber version = {0};
    if (versionComponents.count >= 3) {
        version.majorVersion = [versionComponents[0] integerValue];
        version.minorVersion = [versionComponents[1] integerValue];
        version.patchVersion = [versionComponents[2] integerValue];
    }
    return version;
}

+ (nullable Class<GADAdNetworkExtras>)networkExtrasClass {
    return [AlxAdmobNetworkExtras class];
}

+ (void)setUpWithConfiguration:(GADMediationServerConfiguration *)configuration
             completionHandler:(GADMediationAdapterSetUpCompletionBlock)completionHandler {
    NSLog(@"%@: setUp", TAG);
    
    // This is where you will initialize the SDK that this custom event is built for.
    // Upon finishing the SDK initialization, call the completion handler with success.
    
    if (AlxAdmobBaseAdapter.isInitialized) {
        NSString *errorStr = @"The alx sdk has been initialized";
        NSLog(@"%@: %@", TAG, errorStr);
        completionHandler([NSError errorWithDomain:errorStr code:-100 userInfo:nil]);
        return;
    }
    
    for (GADMediationCredentials * credential in configuration.credentials) {
        NSDictionary<NSString *, id> *result = [AlxAdmobBaseAdapter initSdkFor:credential.settings];
        if ([result[@"success"] boolValue]) {
            completionHandler(nil);
            return;
        }
    }
    completionHandler(nil);
}

- (instancetype)init {
    self = [super init];
    if (self) {
        NSLog(@"%@: init", TAG);
    }
    return self;
}

#pragma mark - Parse Parameter

+ (nullable NSDictionary<NSString *, id> *)parseAdparameterFor:(GADMediationCredentials *)parameters {
    NSString *params = parameters.settings[PARAMETER];
    if (!params) {
        NSString *errorStr = @"The parameter field is not found in the adConfiguration object";
        NSLog(@"%@: %@", TAG, errorStr);
        return nil;
    }
    
    NSData *data = [params dataUsingEncoding:NSUTF8StringEncoding];
    if (!data) {
        NSLog(@"%@: parameter: Data obj is nil", TAG);
        return nil;
    }
    
    NSError *error = nil;
    NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
    if (error) {
        NSLog(@"%@: parameter parse error: %@", TAG, error.localizedDescription);
        return nil;
    }
    
    return json;
}

#pragma mark - SDK Init

+ (NSDictionary<NSString *, id> *)initSdkFor:(nullable NSDictionary<NSString *, id> *)parameters {
    NSLog(@"%@: alx-sdk-version:%@", TAG, [AlxSdk getSDKVersion]);
    NSLog(@"%@: admob-sdk-version:%@", TAG, [self stringForVersionNumber:GADMobileAds.sharedInstance.versionNumber]);
    NSLog(@"%@: admob-adapter-version:%@", TAG, AlxAdmobMetaInfo.ADAPTER_VERSION);
    
    if (!parameters) {
        NSString *errorStr = @"initialize alx params is empty";
        NSLog(@"%@: error: %@", TAG, errorStr);
        return @{@"success": @NO, @"error": errorStr};
    }
    
    // 从 parameters 中获取参数字符串
    NSString *paramsStr = parameters[@"parameter"];
    if (![paramsStr isKindOfClass:[NSString class]] || paramsStr.length == 0) {
        NSString *errorStr = @"parameter string is missing or not a string";
        NSLog(@"%@: error: %@", TAG, errorStr);
        return @{@"success": @NO, @"error": errorStr};
    }
    
    // 将 JSON 字符串转换为 Data
    NSData *admobJSONData = [paramsStr dataUsingEncoding:NSUTF8StringEncoding];
    if (!admobJSONData) {
        NSString *errorStr = @"failed to convert parameter string to data";
        NSLog(@"%@: error: %@", TAG, errorStr);
        return @{@"success": @NO, @"error": errorStr};
    }
    
    // 将 Data 转为 JSON 字典
    NSError *jsonError = nil;
    NSDictionary<NSString *, id> *paramsDict = [NSJSONSerialization JSONObjectWithData:admobJSONData options:0 error:&jsonError];
    if (![paramsDict isKindOfClass:[NSDictionary class]]) {
        NSString *errorStr = @"failed to parse parameter JSON";
        NSLog(@"%@: error: %@", TAG, errorStr);
        return @{@"success": @NO, @"error": errorStr};
    }
    
    NSString *appid = paramsDict[@"appid"];
    NSString *sid = paramsDict[@"sid"];
    NSString *token = paramsDict[@"token"];
    NSString *debug = paramsDict[@"isdebug"];
    
    if (!appid || !sid || !token) {
        NSString *errorStr = @"initialize alx params: appid or sid or token is empty";
        NSLog(@"%@: error: %@", TAG, errorStr);
        return @{@"success": @NO, @"error": errorStr};
    }
    
    NSLog(@"%@: token=%@; appid=%@; sid=%@", TAG, token, appid, sid);
    [AlxSdk initializeSDKWithToken:token sid:sid appId:appid];
    _isInitialized = YES;
    [self sdkInfo];
    
    if (debug && debug.length > 0) {
        if ([debug.lowercaseString isEqualToString:@"true"]) {
            [AlxSdk setDebug:YES];
        } else if ([debug.lowercaseString isEqualToString:@"false"]) {
            [AlxSdk setDebug:NO];
        }
    }
    
    // User Privacy
    // MARK: - GDPR Consent Handling
    NSInteger gdprFlag = [[NSUserDefaults standardUserDefaults] integerForKey:@"IABTCF_gdprApplies"];
    NSString *gdprConsent = [[NSUserDefaults standardUserDefaults] stringForKey:@"IABTCF_TCString"];
    
    // tcf v2 consent
    if (gdprFlag == 1) {
        [AlxSdk setGDPRConsent:YES];
    } else {
        [AlxSdk setGDPRConsent:NO];
    }
    [AlxSdk setGDPRConsentMessage:gdprConsent ?: @""];
    
    return @{@"success": @YES, @"error": @"ok"};
}

+ (void)sdkInfo {
    NSDictionary *data = @{
        @"sdk_name": @"Admob",
        @"sdk_version": [self stringForVersionNumber:GADMobileAds.sharedInstance.versionNumber],
        @"adapter_version": AlxAdmobMetaInfo.ADAPTER_VERSION
    };
    [AlxSdk addExtraParametersWithKey:@"alx_adapter" value:data];
}

+ (NSString *)stringForVersionNumber:(GADVersionNumber)version {
    return [NSString stringWithFormat:@"%ld.%ld.%ld",
            (long)version.majorVersion,
            (long)version.minorVersion,
            (long)version.patchVersion];
}

- (NSError *)errorWithCode:(NSInteger)code msg:(NSString *)msg {
    return [NSError errorWithDomain:TAG
                               code:code
                           userInfo:@{NSLocalizedDescriptionKey: msg}];
}

#pragma mark - Property

+ (BOOL)isInitialized {
    return _isInitialized;
}

+ (void)setIsInitialized:(BOOL)isInitialized {
    _isInitialized = isInitialized;
}

@end

File: custom_adapter/AdmobAdapter/Swift/AlxAdmobBaseAdapter.swift
//
//  AlxAdmobBaseAdapter.swift
//

import Foundation
import GoogleMobileAds
import AlxAds

@objc(AlxAdmobBaseAdapter)
public class AlxAdmobBaseAdapter: NSObject,MediationAdapter {
    
    private static let TAG = "AlxAdmobBaseAdapter"
    private static let PARAMETER = "parameter"
    
    public static var isInitialized:Bool = false
    
    public static func adapterVersion() -> VersionNumber {
        let versionComponents = AlxAdmobMetaInfo.ADAPTER_VERSION.components(
            separatedBy: ".")
        var version = VersionNumber()
        if versionComponents.count >= 3 {
            version.majorVersion = Int(versionComponents[0]) ?? 0
            version.minorVersion = Int(versionComponents[1]) ?? 0
            version.patchVersion = Int(versionComponents[2]) ?? 0
        }
        return version
    }
    
    public static func adSDKVersion() -> VersionNumber {
        let versionComponents = AlxSdk.getSDKVersion().components(separatedBy: ".")
        var version = VersionNumber()
        if versionComponents.count >= 3 {
            version.majorVersion = Int(versionComponents[0]) ?? 0
            version.minorVersion = Int(versionComponents[1]) ?? 0
            version.patchVersion = Int(versionComponents[2]) ?? 0
        }
        return version
    }
    
    public static func networkExtrasClass() -> (any AdNetworkExtras.Type)? {
        return AlxAdmobNetworkExtras.self
    }
    
    public static func setUp(with configuration: MediationServerConfiguration,
                                  completionHandler: GADMediationAdapterSetUpCompletionBlock) {
        NSLog("%@: setUp",TAG)
        // This is where you you will initialize the SDK that this custom event is built for.
        // Upon finishing the SDK initialization, call the completion handler with success.
        
        if AlxAdmobBaseAdapter.isInitialized {
            let errorStr = "The alx sdk has been initialized"
            NSLog("%@: %@",AlxAdmobBaseAdapter.TAG,errorStr)
            completionHandler(NSError(domain: errorStr, code: -100))
            return
        }
        
        for items in configuration.credentials {
            let result = AlxAdmobBaseAdapter.initSdk(for: items.settings)
            if result.success {
                completionHandler(nil)
                return
            }
        }
        completionHandler(nil)
    }
    
    
    required public override init() {
        super.init()
        NSLog("%@: init",AlxAdmobBaseAdapter.TAG)
    }
    
    // MARK: - parase parameter
    public static func parseAdparameter(for parameters: MediationCredentials)-> [String:Any]?{
        guard let params:String = parameters.settings[AlxAdmobBaseAdapter.PARAMETER] as? String else{
            let errorStr="The parameter field is not found in the adConfiguration object"
            NSLog("%@: %@",AlxAdmobBaseAdapter.TAG,errorStr)
            return nil
        }
        
        guard let data = params.data(using: .utf8) else{
            NSLog("%@: parameter: Data obj is nil", AlxAdmobBaseAdapter.TAG)
            return nil
        }
        do{
            let json = try JSONSerialization.jsonObject(with: data,options: []) as! [String:Any]
            return json
        }catch{
            NSLog("%@: parameter parse error: %@", AlxAdmobBaseAdapter.TAG,error.localizedDescription)
        }
        return nil
    }
    
    
    // MARK: - SDK init
    @discardableResult
    public static func initSdk(for parameters: [String:Any]?)->(success:Bool,error:String){
        NSLog("%@: alx-sdk-version:%@",AlxAdmobBaseAdapter.TAG,AlxSdk.getSDKVersion())
        NSLog("%@: admob-sdk-version:%@",AlxAdmobBaseAdapter.TAG,string(for: MobileAds.shared.versionNumber))
        NSLog("%@: admob-adapter-version:%@",AlxAdmobBaseAdapter.TAG,AlxAdmobMetaInfo.ADAPTER_VERSION)
        
        guard let parameters = parameters else {
            let errorStr="initialize alx params is empty"
            NSLog("%@: error: %@",AlxAdmobBaseAdapter.TAG,errorStr)
            return (success:false,error:errorStr)
        }
        
        // 从 parameters 中获取参数字符串 / Get the parameter string from parameters
        guard let paramsStr = parameters["parameter"] as? String else {
            let errorStr = "parameter string is missing or not a string"
            NSLog("%@: error: %@", AlxAdmobBaseAdapter.TAG, errorStr)
            return (success: false, error: errorStr)
        }
        
        // 将 JSON 字符串转换为Data类型 / Convert the JSON string to Data type
        guard let admobJSONData = paramsStr.data(using: .utf8) else {
            let errorStr = "failed to convert parameter string to data"
            NSLog("%@: error: %@", AlxAdmobBaseAdapter.TAG, errorStr)
            return (success: false, error: errorStr)
        }
        
        // 将Data类型转为JSON字典 / Convert Data type to JSON dictionary
        guard let paramsDict = (try? JSONSerialization.jsonObject(with: admobJSONData, options: [])) as? [String: Any] else {
            let errorStr = "failed to parse parameter JSON"
            NSLog("%@: error: %@", AlxAdmobBaseAdapter.TAG, errorStr)
            return (success: false, error: errorStr)
        }
        
        let appid = paramsDict["appid"] as? String
        let sid = paramsDict["sid"] as? String
        let token = paramsDict["token"] as? String
        let debug:String? = paramsDict["isdebug"] as? String
        
        guard let appid=appid,let sid=sid,let token=token else{
            let errorStr="initialize alx params: appid or sid or token is empty"
            NSLog("%@: error: %@",AlxAdmobBaseAdapter.TAG,errorStr)
            return (success:false,error:errorStr)
        }
        
        NSLog("%@: token=%@; appid=%@; sid=%@",AlxAdmobBaseAdapter.TAG,token,appid,sid)
        AlxSdk.initializeSDK(token: token, sid: sid, appId: appid)
        AlxAdmobBaseAdapter.isInitialized = true
        sdkInfo()
        
        if let debug,!debug.isEmpty {
            if debug.lowercased() == "true" {
                AlxSdk.setDebug(true)
            }else if debug.lowercased() == "false" {
                AlxSdk.setDebug(false)
            }
        }
        
        
        // User Privacy
        // MARK: - GDPR Consent Handling
        let gdprFlag = UserDefaults.standard.integer(forKey: "IABTCF_gdprApplies")
        let gdprConsent = UserDefaults.standard.string(forKey: "IABTCF_TCString")
        // tcf v2 consent
        if gdprFlag == 1{
            AlxSdk.setGDPRConsent(true)
        }else  {
            AlxSdk.setGDPRConsent(false)
        }
        AlxSdk.setGDPRConsentMessage(gdprConsent ?? "")
        
        return (success:true,error:"ok")
    }
    
    
    
    private static func sdkInfo(){
        var data:[String:String] = [:]
        data["sdk_name"] = "Admob"
        data["sdk_version"] = string(for: MobileAds.shared.versionNumber)
        data["adapter_version"] = AlxAdmobMetaInfo.ADAPTER_VERSION
        AlxSdk.addExtraParameters(key:"alx_adapter",value:data)
    }
    
    public func error(code:Int,msg:String) -> NSError{
        return NSError(domain: "\(AlxAdmobBaseAdapter.TAG)", code: code, userInfo: [NSLocalizedDescriptionKey : msg])
    }
    
    

}
